extras/FvwmCommand/FvwmCommand.1
.\" t
.\" @(#)FvwmCommand.1	7/19/97
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.25i
..
.de EE
.fi
.in -.25i
.if n .sp 1
.if t .sp .5
..
.ta .2i .4i .6i .8i 1i 1.2i 1.4i
.TH FvwmCommand 1x "July 19 1997"
.UC
.SH NAME
FvwmCommand \- FVWM command external interface
.SH SYNOPSIS
In .fvwm2rc file:

.EX 
Module FvwmCommandS [name]
.EE

Then, in script file or from shell:

.EX
FvwmCommand [-mrvw] [-i level] [-f name] [command...]
.EE

.SH DESCRIPTION

FvwmCommand lets you monitor fvwm transaction and issue fvwm command 
from a shell command line or scripts. 
FvwmCommand takes each argument as a fvwm command. Quotes can be 
used to send commands including spaces.

.br
FvwmCommand.pm is for perl scripts in order 
to keep the syntax similar to fvwm configuration file. 

.EX
use FvwmCommand;
if( $ARGV[0] eq 'home' ) { 
    Desk 0,0; GotoPage '1 1'; 
}elsif( $ARGV[0] eq 'jump' ) { 
    Desk "0 2"; GotoPage 1, 1; 
}
.EE

Commas can be used to separate Fvwm commands' arguments or quotes
can be used combine them. The commands are case sensitive.

.br
FvwmCommand.sh is collection of functions for bourn shell or its
cousines. It is to be sourced:

.EX
\&. FvwmCommand.sh
.br
DesktopSize 5x5
.EE

The commands in FvwmCommand.sh and FvwmCommand.pl are case sensitive. 


.SH INVOCATION
FvwmCommandS should be spawned once by fvwm, either in .fvwm2rc file,
from menu, or from FvwmConsole.
From then on, FvwmCommand
can be called from a shell or script to execute fvwm commands.
.SH OPTIONS
.IP "\fI-f <name>\fR" 0.4i
Specifies an alternative fifo set to communicate with FvwmCommandS.
The default fifo set is ~/.FvwmCommand, in which
~/.FvwmCommandC is used to send commands and ~/.FvwmCommandM is to receive
messages.
FvwmCommandS must have been invoked with the same <name> as its first argument
prior to FvwmCommand invokation.
This option is useful when a dedicated connection is necessary to run a background
job while another connection is kept for interactive use.

.IP "\fI-i <level>\fR"
Specified the level of information FvwmCommand outputs.
.sp
.RS
.IP 0 0.4i
Error messages only. Note that Fvwm doesn't return any error messages in
cases like below since 'windowid' itself is a valid command.
.sp
.EX
FvwmCommand -i 0 'windowid abcd move'
.EE
.IP 1
Errors and window configuration information. This is the default.
.EX
FvwmCommand FvwmIconMan 
.EE
The following infomation will show up.
.EX
0x03400003 window               FvwmIconMan
0x03400003 icon                 FvwmIconMan
0x03400003 class                FvwmModule
0x03400003 resource             FvwmIconMan
.EE
The first column shows the window ID number, which can be used in 'windowid' command.
The second column shows the information types. 
The last column shows the information contents.
If no information is returned, add -w <time> option. This might be needed in heavily
loaded systems.
.IP 2
Above and static window information.
.EX
FvwmCommand -i2 'FvwmIconBox' 
.EE
The below is its output.
.EX
0x02400002 frame                x 0, y 0, width 410, height 116
0x02400002 desktop              0
0x02400002 StartIconic          no
0x02400002 OnTop                no
0x02400002 Sticky               yes
0x02400002 WindowListSkip       yes
0x02400002 SuppressIcon         no
0x02400002 NoiconTitle          no
0x02400002 Lenience             no
0x02400002 StickyIcon           no
0x02400002 CirculateSkipIcon    no
0x02400002 CirculateSkip        no
0x02400002 ClickToFocus         no
0x02400002 SloppyFocus          no
0x02400002 SkipMapping          no
0x02400002 Handles              no
0x02400002 Title                no
0x02400002 Mapped               no
0x02400002 Iconified            no
0x02400002 Transient            no
0x02400002 Raised               no
0x02400002 Visible              no
0x02400002 IconOurs             no
0x02400002 PixmapOurs           no
0x02400002 ShapedIcon           no
0x02400002 Maximized            no
0x02400002 WmTakeFocus          no
0x02400002 WmDeleteWindow       yes
0x02400002 IconMoved            no
0x02400002 IconUnmapped         no
0x02400002 MapPending           no
0x02400002 HintOverride         yes
0x02400002 MWMButtons           no
0x02400002 MWMBorders           no
0x02400002 title height         0
0x02400002 border width         4
0x02400002 base size            width 0, height 0
0x02400002 size increment       width 72, height 66
0x02400002 min size             width 114, height 108
0x02400002 max size             width 32767, height 32767
0x02400002 gravity              NorthWest
0x02400002 pixel                text 0xffff, back 0x7bef
0x02400002 window               FvwmIconBox
0x02400002 icon                 FvwmIconBox
0x02400002 class                FvwmIconBox
0x02400002 resource             FvwmIconBox
.EE
.IP 3
All information available.
.EX
FvwmCommand -i3 'Killmodule FvwmIcon*'    
.EE
This will report which windows are closed.
.EX
0x03400003 destroy
0x02400002 destroy
.EE
.RE

.IP "\fI-m\fR"
Monitors fvwm window information transaction. FvwmCommand continuously outputs
information it receives without exiting. This is used in a
background job often combined with -i3 option in order to control windows
dynamically.
.EX
FvwmCommand -mi3 | grep 'iconify'
.EE
It will report when windows are iconified or deiconified.
.sp
Note: FvwmCommand does not buffer its stdout but many utilities such as 
grep or sed uses block buffer. The output of the next example will not show up
until either FvwmCommand is terminated or stdout buffer from
grep is filled.
.EX
FvwmCommand -mi3 | grep ' map' | 
sed 's/\(0x[0-9a-f]*\).*/windowid \1 move 0 0/'
.EE
Instead, use tools with buffer control such as perl. The below will iconify 
new windows when opened.
.EX
Fvwm -mi3 | perl -ne '
$|=1; 
print "windowid $1 iconify\\n" if /^(0x\\S+) add/;
\' > ~/\.FvwmCommandC
.EE
.IP "\fI-r\fR"
Waits for a reply before it exits. 
FvwmCommand exits if no information or error is returned in a fixed amount of 
time period.
This option overrides this time limit and wait for at least one message
back. 
After the initial message, it will wait for another message for the time
limit.
This option is useful when the system is too load to make any predition
when the system is responding AND the command causes some
message to be sent back.

.IP "\fI-v\fR"
Returns FvwmCommand version number and exits.

.IP "\fI-w <time>\fR"
Waits for <time> micro seconds for a message.
FvwmCommand exits if no information or error is returned in a fixed amount of 
time period. The default is 500 ms. This option overrides this default value.
.EX
FvwmCommand -w 0 'killme'
.EE

.SH ERRORS
If the following error message show up, it is most likely that FvwmCommandS
is not running.
.EX
FvwmCommand error in opening message
--No such file or directory--
.EE

.SH SEE ALSO
fvwm2

.SH AUTHOR
Toshi Isogai  isogai@enshin.com




NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

extras/FvwmCommand/FvwmCommand.c
/* $Id$
 * $Source$
 *
 * Fvwm2 command input interface.
 *
 * Copyright 1998, Toshi Isogai.
 * Use this program at your own risk. 
 * Permission to use this program for any purpose is given,
 * as long as the copyright is kept intact. 
 *
 */

#include "FvwmCommand.h"

#define MYNAME "FvwmCommand"

int  Fdr, Fdw;  /* file discriptor for fifo */
FILE *Frun;     /* File contains pid */
char *Fr_name;  
int  Pfd;
char *getline();
fd_set fdset;

struct timeval Tv;
int  Opt_reply; /* wait for replay */
int  Opt_monitor;     
int  Opt_info;
int  Opt_Serv;
int  Opt_flags;
FILE *Fp;
int  Rc;  /* return code */
int  Bg;  /* FvwmCommand in background */

void err_msg( char *msg ); 
void err_quit( char *msg ); 
void sendit( char *cmd );
void receive( void );
void sig_ttin ( int );
void sig_pipe ( int );
void sig_quit ( int );
void usage(void);
int  read_f (int fd, char *p, int len);
void close_fifos (void);

void process_message( void ) ;
void list( unsigned long *body, char *) ;
void list_configure(unsigned long *body);
void list_focus_change(unsigned long *body) ;
void list_header(unsigned long *body, char *) ;
void list_icon_loc(unsigned long *body) ;
void list_iconify(unsigned long *body);
void list_mini_icon(unsigned long *body) ;
void list_new_desk(unsigned long *body) ;
void list_new_page(unsigned long *body) ;
void list_string (char *str);
void spawn_child( void );


/*******************************************************
 *
 * send command to and receive message from the server 
 *
 *******************************************************/
int main ( int argc, char *argv[]) {
  char cmd[MAX_COMMAND_SIZE];
  char *home;
  char *f_stem, *fc_name, *fm_name;
  char *sf_stem;
  int  i;
  int  opt;
  int  ncnt;
  int  count;
  struct timeval tv2;
  extern char *optarg;
  extern int  optind, opterr, optopt;
  
  signal (SIGINT, sig_quit);  
  signal (SIGHUP, sig_quit);  
  signal (SIGQUIT, sig_quit);  
  signal (SIGTERM, sig_quit);  
  signal (SIGTTIN, sig_ttin);
  signal (SIGTTOU, sig_ttin);

  Opt_reply = 0;
  Opt_info = 1;
  f_stem = NULL;
  sf_stem = NULL;
  Opt_monitor = 0;
  Opt_Serv = 0;
  Opt_flags = 2;
  Tv.tv_sec = 0;
  Tv.tv_usec = 500000;
  Rc = 0;
  Bg = 0;


  while( (opt = getopt( argc, argv, "S:hvF:f:w:i:rm" )) != EOF ) {
    switch(opt) {
    case 'h':
      usage();
      exit(0);
      break;
    case 'f': 
      f_stem = optarg;
      break;
    case 'F': 
      Opt_flags = atoi (optarg);
      break;
    case 'S':
      sf_stem = optarg;
      Opt_Serv = 1;
      break;
    case 'i':
      Opt_info = atoi( optarg );
      break;
    case 'v':
      printf("%s %s\n", MYNAME, MYVERSION );
      exit(0);
    case 'w':
      Tv.tv_usec = atoi( optarg ) % 1000000;
      Tv.tv_sec = atoi( optarg ) / 1000000;
      break;
    case 'm':
      Opt_monitor = 1;
      break;
    case 'r':
      Opt_reply = 1;
      break;
    case '?':
      exit(3);
    }
  }


  if( f_stem == NULL ) {
    home = getenv("HOME");
    f_stem = safemalloc( strlen(home)+ strlen(F_NAME) + 3);
    strcpy (f_stem, home);
    if (f_stem[strlen(f_stem)-1] != '/') {
      strcat (f_stem, "/");
    }
    strcat (f_stem, F_NAME);
  }

  /* create 2 fifos */
  fm_name = safemalloc( strlen(f_stem) + 2 );
  strcpy(fm_name,f_stem);
  strcat(fm_name, "M");
  fc_name = safemalloc( strlen(f_stem) + 2 );
  strcpy(fc_name,f_stem);
  strcat(fc_name, "C");
  Fr_name = safemalloc( strlen(f_stem) + 2 );
  strcpy(Fr_name,f_stem);
  strcat(Fr_name, "R");

  if ((Frun = fopen (Fr_name,"r" )) !=NULL) {
    if (fgets (cmd, 20, Frun) != NULL) {
      fprintf (stderr, "\nFvwmCommand lock file  %sR is detected. "
	       "This may indicate another FvwmCommand is running. "
	       "It appears to be running under process ID:\n%s\n", 
	       f_stem, cmd );
      fprintf (stderr, "You may either kill the process or run FvwmCommand "
	       "with another FIFO set using option -S and -f. "
	       "If the process doesn't exist, simply remove file:\n%sR\n\n", 
	       f_stem);
      exit(1);
    }
    fclose (Frun);
    unlink (Fr_name);
  }
	       
  if( Opt_Serv ) {
    sprintf (cmd,"%s '%sS %s'", argv[0], MYNAME, sf_stem);
    system (cmd);
  }

  if ((Frun = fopen (Fr_name,"w" )) != NUll) {
    fprintf (Frun, "%d\n", (int) getpid());
    fclose (Frun);
  }else {
      err_quit ("writing lock file");
  }
    
  Fdr = Fdw = -1;
  count = 0;
  while ((Fdr=open (fm_name, O_RDONLY)) < 0) {
    if (count++>5) {
      err_quit ("opening message fifo");
    }
    sleep(1);
  }
  count = 0;
  while ((Fdw=open (fc_name, O_WRONLY)) < 0) {
    if (count++>2) {
      err_quit ("opening command fifo");
    }
    sleep(1);
  }

  strcpy (cmd, CMD_CONNECT);
  sendit (cmd);

  i = optind;
  if( Opt_monitor ) {

    /* test if its stdin is closed for coprocess */
    tv2.tv_sec = 0;
    tv2.tv_usec = 5;
    FD_ZERO(&fdset);
    FD_SET(STDIN_FILENO, &fdset); 
    ncnt = _select(FD_SETSIZE,&fdset, 0, 0, &tv2);
    if( ncnt && (fgets( cmd, 1, stdin )==0 || cmd[0] == 0)) {
      Bg = 1;
    }

    /* line buffer stdout for coprocess */
    setvbuf( stdout, NULL, _IOLBF, 0); 

    /* send arguments first */
    for( ;i < argc; i++ ) {
      strncpy( cmd, argv[i], MAX_COMMAND_SIZE-2 );
      sendit( cmd );
    }
		

    while(1) {
      FD_ZERO(&fdset);
      FD_SET(Fdr, &fdset);
      if( Bg == 0 ) {
	FD_SET(STDIN_FILENO, &fdset); 
      }
      ncnt = _select(FD_SETSIZE,&fdset, 0, 0, NULL);
			
      /* message from fvwm */
      if (FD_ISSET(Fdr, &fdset)){
	process_message();
      }
      
      if( Bg == 0 ) {
      /* command input */
	if( FD_ISSET(STDIN_FILENO, &fdset) ) {
	  if( fgets( cmd, MAX_COMMAND_SIZE-2, stdin ) == 0 ) {
	    if( Bg == 0 ) {
	      /* other than SIGTTIN */
	      break;
	    }else{
	      continue;
	    }
	  }
	  sendit( cmd );
	}
      }
    }
  }else {
    for( ;i < argc; i++ ) {
      strncpy( cmd, argv[i], MAX_COMMAND_SIZE-2 );
      sendit( cmd );
      receive();
    }
  }
  close_fifos();
  exit( Rc );
}

/*
 * send exit notice and close fifos
 */
void close_fifos (void) {
  char cmd[10];

  if (Fdw >= 0) {
    strcpy (cmd, CMD_EXIT);
    sendit (cmd);
  }
  close(Fdr);
  close(Fdw);
  unlink (Fr_name);
}

/******************************************
 *  signal handlers
 ******************************************/
void sig_quit (int dummy) {
  close_fifos();
  err_msg("receiving signal\n" );
  exit(1);
}

void sig_ttin( int  dummy ) {
  Bg = 1;
  signal( SIGTTIN, SIG_IGN ); 
}

/************************************/
/* print error message on stderr */
/************************************/
void err_quit( char *msg ) {
  fprintf (stderr, "%s ", strerror(errno));
  err_msg(msg);
  close_fifos();
  exit(1);
}

void err_msg( char *msg ) {
  fprintf( stderr, "%s error in %s\n", MYNAME , msg );
}

/*************************************/
/* add cr to the command and send it */
/*************************************/
void sendit( char *cmd ) {
  int clen;
	
  if( cmd[0] != '\0'  ) {
    clen = strlen(cmd);
		
    /* add cr */
    if( cmd[clen-1] != '\n' ) {
      strcat(cmd, "\n");
      clen++;
    }
    if( clen != 1 ) {
      write( Fdw, cmd, clen );
    }
  }
}

void receive () {
  int  ncnt;
  struct timeval tv;

  if( Opt_reply ) {
    /* wait indefinitely */
    FD_ZERO(&fdset);
    FD_SET( Fdr, &fdset);
    ncnt = _select(FD_SETSIZE,&fdset, 0, 0, NULL);
  }
		

  while (1){
    tv.tv_sec = Tv.tv_sec;
    tv.tv_usec = Tv.tv_usec;
    FD_ZERO(&fdset);
    FD_SET(Fdr, &fdset);
    
    ncnt = _select(FD_SETSIZE,&fdset, 0, 0, &tv);
    
    if( ncnt < 0 ) {
      err_quit("receive");
      break;
    }
    if( ncnt == 0 ) {
      /* timeout */
      break;
    }
    if (FD_ISSET(Fdr, &fdset)) {
      process_message();
    }
  }
}


/*******************************
 * print usage
 *******************************/
void usage(void) {
  fprintf (stderr, "Usage: %s [OPTION] [COMMAND]...\n", MYNAME);
  fprintf (stderr, "Send commands to fvwm2 via %sS\n\n", MYNAME);
  fprintf (stderr, 
	   "  -F <flag info>      0 - no flag info\n");
  fprintf (stderr, 
	   "                      2 - full flag info (default)\n");
  fprintf (stderr, 
	   "  -S <file name>      "
	   "invoke another %s server with fifo <file name>\n",
	   MYNAME);
  fprintf (stderr, 
	   "  -f <file name>      use fifo <file name> to connect to %sS\n", 
	   MYNAME);
  fprintf (stderr, 
	   "  -i <info level>     0 - error only\n" );
  fprintf (stderr, 
	   "                      1 - above and config info (default)\n" );
  fprintf (stderr, 
	   "                      2 - above and static info\n" );
  fprintf (stderr, 
	   "                      3 - above and dynamic info\n" );
  fprintf (stderr, 
	   "  -m                  monitor fvwm2 message transaction\n");
  fprintf (stderr, 
	   "  -r                  "
	   "wait for a reply (overrides waiting time)\n");
  fprintf (stderr, 
	   "  -v                  print version number\n");
  fprintf (stderr, 
	   "  -w <micro sec>      waiting time for the reponse from fvwm\n");
  fprintf (stderr, "\nDefault fifo names are ~/.%sC and ~/.%sM\n",
	   MYNAME, MYNAME);
  fprintf (stderr, "Default waiting time is 500,000 us\n");
}
 
/*
 * read fifo
 */
int read_f (int fd, char *p, int len) {
  int i, n;
  for (i=0; i<len; )  {
    n = read (fd, &p[i], len-i);
    if (n<0 && errno!=EAGAIN) {
      err_quit("reading message");
    }
    if (n==0) {
      /* eof */
      close_fifos();
      exit( Rc );
    }
    i += n;
  }
  return len;
}

/*
 * process message
 */
void process_message( void ) {
  unsigned long type, length, body[24*SOL];
	
  read_f( Fdr, (char*)&type, SOL);
  read_f( Fdr, (char*)&length, SOL);
  read_f( Fdr, (char*)body, length );

  if( type==M_ERROR ) {
    fprintf( stderr,"%s", (char *)&body[3] );
    Rc = 1;
  }else if( Opt_info >= 1 ) {

    switch( type ) {

    case M_WINDOW_NAME:
      list( body, "window");
      break;
    case M_ICON_NAME:
      list(body, "icon");
      break;
    case M_RES_CLASS:
      list(body, "class" );
      break;
    case M_RES_NAME:
      list(body, "resource");
      break;
					
    case M_END_WINDOWLIST:
      list_string("end windowlist");
      break;
					
    case M_ICON_FILE:
      list(body, "icon file");
      break;
    case M_ICON_LOCATION:
      list_icon_loc(body);
      break;
					
    case M_END_CONFIG_INFO:
      list_string("end configinfo");
      break;
			
    case M_DEFAULTICON:
      list(body, "default icon");
      break;
			
    case M_MINI_ICON:
      list_mini_icon( body );
      break;
			
    case M_CONFIG_INFO:
      printf( "%s", (char *)&body[3] );
      break;
				
    default:
      if( Opt_info >=2 ) {
				
	switch(type) {
			
	case M_CONFIGURE_WINDOW:
	  list_configure( body);
	  break;
	case M_STRING:
	  list(body, "string");
	  break;
			
	default:
	  if( Opt_info >= 3 ) {
	    switch( type ) {
	    case M_NEW_PAGE:
	      list_new_page(body);
	      break;
	    case M_NEW_DESK:
	      list_new_desk(body);
	      break;
	    case M_ADD_WINDOW:
	      list_header(body, "add");
	      list_configure( body);
	      break;
	    case M_RAISE_WINDOW:
	      list_header(body, "raise");
	      break;
	    case M_LOWER_WINDOW:
	      list_header(body, "lower");
	      break;
	    case M_FOCUS_CHANGE:
	      list_focus_change( body );
	      break;
	    case M_DESTROY_WINDOW:
	      list_header(body, "destroy");
	      break;
	    case M_ICONIFY:
	      list_iconify( body );
	      break;
	    case M_DEICONIFY:
	      list_header(body, "deiconify");
	      break;
	    case M_MAP:
	      list_header(body, "map");
	      break;
	    case M_WINDOWSHADE:
	      list_header(body, "windowshade");
	      break;
	    case M_DEWINDOWSHADE:
	      list_header(body, "dewindowshade");
	      break;
	    default:
	      printf("0x%lx type 0x%lx\n", body[0], type );
	    }
	  }
	}
      }
    }
  }
}

        			
/**********************************
 *
 *  print configuration info
 *
 **********************************/

void list_configure(unsigned long *body) {
  unsigned long i;
  char *flag;
  char *grav;
  char *flagstr[32] = {
    "StartIconic",
    "OnTop",
    "Sticky",
    "WindowListSkip",
    "SuppressIcon",
    "NoiconTitle",
    "Lenience",
    "StickyIcon",
    "CirculateSkipIcon",
    "CirculateSkip",
    "ClickToFocus",
    "SloppyFocus",
    "SkipMapping",
    "Handles",
    "Title",
    "Mapped",
    "Iconified",
    "Transient",
    "Raised", /* ??? */
    "Visible",
    "IconOurs",
    "PixmapOurs",
    "ShapedIcon",
    "Maximized",
    "WmTakeFocus",
    "WmDeleteWindow",
    "IconMoved",
    "IconUnmapped",
    "MapPending",
    "HintOverride",
    "MWMButtons",
    "MWMBorders"
  };


  printf( "0x%08lx %-20s x %ld, y %ld, width %ld, height %ld\n",
	  body[0], "frame", body[3], body[4], body[5], body[6] );
  printf( "0x%08lx %-20s %ld\n" ,body[0], "desktop", body[7]);

  if (Opt_flags == 2) {
    for( i=0; i<=31; i++ ) {
      if( body[8] & (1<<i) ) {
	flag = "yes";
      }else{
	flag = "no";
      }
      printf( "0x%08lx %-20s %s\n", body[0], flagstr[i], flag );
    }
  }

  printf( "0x%08lx %-20s %ld\n",
	  body[0], "title height", body[9]);
  printf( "0x%08lx %-20s %ld\n",
	  body[0], "border width", body[10]);
  printf( "0x%08lx %-20s width %ld, height %ld\n",
	  body[0], "base size", body[11], body[12]);
  printf( "0x%08lx %-20s width %ld, height %ld\n",
	  body[0], "size increment", body[13], body[14]);
  printf( "0x%08lx %-20s width %ld, height %ld\n",
	  body[0], "min size", body[15], body[16]);
  printf( "0x%08lx %-20s width %ld, height %ld\n",
	  body[0], "max size", body[17], body[18]);
	

  switch(body[21]) {
  case ForgetGravity:
    grav = "Forget";
    break;
  case NorthWestGravity:
    grav = "NorthWest";
    break;
  case NorthGravity:
    grav = "North";
    break;
  case NorthEastGravity:
    grav = "NorthEast";
    break;
  case WestGravity:
    grav = "West";
    break;
  case CenterGravity:
    grav = "Center";
    break;
  case EastGravity:
    grav = "East";
    break;
  case SouthWestGravity:
    grav = "SouthWest";
    break;
  case SouthGravity:
    grav = "South";
    break;
  case SouthEastGravity:
    grav = "SouthEast";
    break;
  case StaticGravity:
    grav = "Static";
    break;
  default:
    grav = "Unknown";
    break;
  }
  printf( "0x%08lx %-20s %s\n", body[0], "gravity", grav);

  printf( "0x%08lx %-20s text 0x%lx, back 0x%lx\n",
	  body[0], "pixel", body[22], body[23]);
}

/*************************************************************************
 *
 * print info  icon location
 *
 ************************************************************************/
void list_icon_loc(unsigned long *body) {

  printf( "0x%08lx %-20s x %ld, y %ld, width %ld, height%ld\n",
	  body[0], "icon location", body[3], body[4], body[5], body[6] );
}

/*************************************************************************
 *
 * print info mini icon
 *
 ************************************************************************/
void list_mini_icon(unsigned long *body) {

  printf( "0x%08lx %-20s width %ld, height %ld, depth %ld\n",
	  body[0], "mini icon",body[5], body[6], body[7] );
}

/*************************************************************************
 *
 * print info  message body[3] 
 *
 ************************************************************************/
void list( unsigned long *body, char *text ) {

  printf( "0x%08lx %-20s %s\n", body[0], text, (char *)&body[3] );
}


/*************************************************************************
 *
 * print info  new page
 *
 ************************************************************************/
void list_new_page(unsigned long *body) {

  printf( "           %-20s x %ld, y %ld, desk %ld, max x %ld, max y %ld\n",
	  "new page", 
	  body[0], body[0], body[2], body[3], body[4]);
}

/*************************************************************************
 *
 * print info  new desk
 *
 ************************************************************************/
void list_new_desk(unsigned long *body) {
  
  printf( "           %-20s %ld\n",  "new desk", body[0] );
}

/*************************************************************************
 *
 * print string
 *
 ************************************************************************/
void list_string (char *str) {
    printf( "%-20s\n", str );
}

/*************************************************************************
 *
 * print info header
 *
 ************************************************************************/
void list_header(unsigned long *body, char *text) {

  printf("0x%08lx %s\n", body[0], text); 

}

/*************************************************************************
 *
 * print info focus change
 *
 ************************************************************************/
void list_focus_change(unsigned long *body) {

  printf( "0x%08lx %-20s highlight 0x%lx, foreground 0x%lx, background 0x%lx\n",
	  body[0], "focus change", body[3], body[4], body[5] );
}

/*************************************************************************
 *
 * print info iconify
 *
 ************************************************************************/
void list_iconify(unsigned long *body) {

  printf( "0x%08lx %-20s x %ld, y %ld, width %ld, hight %ld\n",
	  body[0], "iconify", body[3], body[4], body[5], body[6] );
}

